packages/@glimmer/runtime/lib/destroyables.ts

```diff
@@ -0,0 +1,187 @@
+import { DEBUG } from '@glimmer/env';
+import { Destroyable, Destructor } from '@glimmer/interfaces';
+import { _WeakSet as WeakSet } from '@glimmer/util';
+
+const EAGER_DESTRUCTORS = new WeakMap<Destroyable, Destructor<Destroyable>[]>();
```

Comment:
way too many weak collections, use one weak collection to associate the monomorphic destroyable state.  also, optimize if things will often have 1 to 1 maybe consider typing it `Destroyable | Destroyable[]`